src/compiler/checker.ts

```diff
@@ -4391,22 +4400,28 @@ namespace ts {
                     // Use explicitly specified property name ({ p: xxx } form), or otherwise the implied name ({ p } form)
                     const name = declaration.propertyName || <Identifier>declaration.name;
                     if (isComputedNonLiteralName(name)) {
-                        // computed properties with non-literal names are treated as 'any'
-                        return anyType;
+                        type = checkComputedDestructuredProperty(parentType, name, text => {

```

Comment:
I'd write a named function in scope for this lambda and then reuse it below in the `else` branch.